test/hotspot/jtreg/serviceability/sa/UniqueVtableTest.java

```diff
@@ -0,0 +1,177 @@
+ * questions.
+ */
+
+/**
+ * @test
+ *
+ * @library /test/lib
+ * @requires vm.hasSA
+ * @modules jdk.hotspot.agent/sun.jvm.hotspot
+ *          jdk.hotspot.agent/sun.jvm.hotspot.debugger
+ *          jdk.hotspot.agent/sun.jvm.hotspot.types
+ *          jdk.hotspot.agent/sun.jvm.hotspot.types.basic
+ *
+ * @run main/othervm --add-opens=jdk.hotspot.agent/sun.jvm.hotspot.types.basic=ALL-UNNAMED UniqueVtableTest
+ */
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import sun.jvm.hotspot.HotSpotAgent;
+import sun.jvm.hotspot.debugger.Address;
+import sun.jvm.hotspot.types.Type;
+import sun.jvm.hotspot.types.basic.BasicTypeDataBase;
+
+import jdk.test.lib.apps.LingeredApp;
+import jdk.test.lib.SA.SATestUtils;
+
+
+public class UniqueVtableTest {
+
+    private HotSpotAgent agent;
+    private MethodHandle vtblForType;
+
+    public UniqueVtableTest() {
+    }
+
+    private static String type2String(Type t) {
+        return t + " (extends " + t.getSuperclass() + ")";
+    }
+
+    private static void log(Object o) {
+        System.out.println(o);
+    }
+
+    private void attach(long pid) throws Throwable {
+        agent = new HotSpotAgent();
+        log("Attaching to process ID " + pid + "...");
+        agent.attach((int) pid);
+        log("Attached successfully.");
+
+        // agent.getTypeDataBase() returns HotSpotTypeDataBase (extends BasicTypeDataBase)
+        // We need a method from BasicTypeDataBase
+        //    Address vtblForType(Type type);
+
+        MethodHandles.Lookup lookup = MethodHandles.lookup();
+        MethodHandles.Lookup classLookup = MethodHandles.privateLookupIn(BasicTypeDataBase.class, lookup);
+        vtblForType = classLookup.findVirtual(BasicTypeDataBase.class, "vtblForType",
+                 MethodType.methodType(Address.class, Type.class));
+    }
+
+    private void detach() {
+        if (agent != null) {
+            agent.detach();
+        }
+    }
+
+    private void runTest() throws Throwable {
+        Map<Address, List<Type>> types = new HashMap<>();
+        Iterator<Type> it = agent.getTypeDataBase().getTypes();
+        int dupFound = 0;
+        while (it.hasNext()) {
+            Type t = it.next();
+            Address vtable = (Address) vtblForType.invoke(agent.getTypeDataBase(), t);
+
+            if (vtable != null) {
+                List<Type> typeList = types.get(vtable);
+                if (typeList == null) {
+                    types.put(vtable, new ArrayList<>(List.of(t)));
+                } else {
+                    // duplicate found
+                    dupFound++;
+                    typeList.add(t);
+                }
+            }
+
+            if (vtable == null && t.getSuperclass() != null) {
```

Comment:
Why only log if there is no superclass?

Reply:
vmStruct contains a lot of entries which should have super == null
I added logging of some stats to the test, on win-x64-fastdebug it reports:
total: 861, no vtable: 503, no_vtable_with_super: 24
if test would log everything with no vtable, test log is truncated

Reply:
In that case is it worth listing the "no vtable" classes that have a superclass? Are they somehow more interesting than those without a superclass?

Reply:
If superclass is set then it's the type which describes VM class (but some JM classes have super == null)
Maybe the test should report all Types with CInt: false, CStr: false, JPrimitive: false, Oop: false, Ptr: false
Looks like these entries correspond declare_toplevel_type/declare_type in vmStruct

Reply:
Updated criteria to log classes without vtable